### ISSUE TITLE
Eager-mode grad of matmul materializes the full M×K×N intermediate (8 GiB allocation at [2048,256] @ [256,4096])

### ISSUE THREAD

**Issue by NeoVand** [ekzhang/jax-js#150]
Found while benchmarking a small GPT training loop (`@jax-js/jax` 0.1.18, webgpu backend, M4 Mac, Chromium 148).

In eager mode, taking a gradient through a plain matmul seems to materialize the whole M×K×N product before reducing, instead of lowering dW to `Xᵀ @ dY`:

```js
// after init() with the webgpu backend
const x = np.ones([2048, 256]);
const w = np.ones([256, 4096]);
const g = grad((ww) => np.sum(np.dot(x.ref, ww)))(w);
await blockUntilReady(g);
```

This tries to allocate an 8 GiB buffer (2048·256·4096 f32 = 2³¹ elements) and the device errors out:

```
Buffer size (8589934592) exceeds the max buffer size limit (4294967292).
 - While calling [Device].CreateBuffer([BufferDescriptor]).
```

Smaller shapes survive but are slow in a way that's consistent with materialize-then-reduce (timing the grad call after warmup, avg of 5):

- 2048×256×512: ~76 ms — the equivalent matmul would be well under 1 ms at the ~1.1 TFLOP/s I measure for forward matmuls on this GPU
- 2048×256×1024: ~182 ms
- 2048×256×4096: the device error above

Wrapping the same computation in `jit` avoids all of this — my full training step with these exact shapes inside runs fine jitted — so it looks specific to the eager path rather than anything fundamental.

Not a blocker for me since I can jit everything, but it's a rough edge: an eager training loop (quectoGPT-style) dies on it as soon as vocab and batch·seq get realistic, and the failure shows up as an async device error far from the offending op. If the fix is "express dot's transpose as dot" somewhere central, I'm happy to attempt a PR with some guidance.

**Comment by ekzhang** [comment]
Thanks! This is definitely not intended behavior, we should have a test
